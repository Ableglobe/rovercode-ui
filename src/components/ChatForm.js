import React from 'react';
import axios from 'axios';
import PropTypes from 'prop-types';
import { hot } from 'react-hot-loader';
import { connect } from 'react-redux';
import { withCookies, Cookies } from 'react-cookie';
import {
  Button, Card, Dropdown, Input,
} from 'semantic-ui-react';
import {
  categorySelectChange as actionCategorySelectChange,
  bodyInputChange as actionBodyInputChange,
  subjectInputChange as actionSubjectInputChange,
  experienceSelectChange as actionExperienceSelectChange,
  WHAT_NEXT, FUN_IDEAS, DEBUG,
  EXPERT, PRETTY_NEW, FIRST_TIME, FOR_A_WHILE,
} from '../actions/chatform';

const mapStateToProps = ({ chatform }) => ({ ...chatform });
const mapDispatchToProps = dispatch => ({
  categorySelectChange: state => dispatch(actionCategorySelectChange(state)),
  bodyInputChange: state => dispatch(actionBodyInputChange(state)),
  subjectInputChange: state => dispatch(actionSubjectInputChange(state)),
  experienceSelectChange: state => dispatch(actionExperienceSelectChange(state)),
});

const helpCategories = [
  {
    value: DEBUG,
    text: "My code isn't doing what I think it should",
  },
  {
    value: WHAT_NEXT,
    text: "I don't know what to do next",
  },
  {
    value: FUN_IDEAS,
    text: "I'm looking for fun ideas",
  },
];

const experience = [
  {
    value: FIRST_TIME,
    text: 'This is my first time on Rovercode',
  },
  {
    value: PRETTY_NEW,
    text: "I'm pretty new to Rovercode",
  },
  {
    value: FOR_A_WHILE,
    text: "I've used Rovercode for a while",
  },
  {
    value: EXPERT,
    text: "I'm a Rovercode expert",
  },
];

class ChatForm extends React.Component {
    // action creators
    categorySelectChange = (event, data) => {
      const payload = data.value;
      const { categorySelectChange } = this.props;
      console.log(payload);
      categorySelectChange(payload);
    };

    subjectInputChange = (event, data) => {
      const payload = data.value;
      const { subjectInputChange } = this.props;
      console.log(payload);
      subjectInputChange(payload);
    };

    bodyInputChange = (event, data) => {
      const payload = data.value;
      const { bodyInputChange } = this.props;
      console.log(payload);
      bodyInputChange(payload);
    };

    experienceSelectChange = (event, data) => {
      const payload = data.value;
      const { experienceSelectChange } = this.props;
      console.log(payload);
      experienceSelectChange(payload);
    };

    validateForm = (experienceValue, categoryValue, subjectValue, bodyValue) => {
      if (experienceValue !== ''
          && subjectValue !== ''
          && bodyValue !== '') {
        this.postData(
          {
            subject: subjectValue,
            body: bodyValue,
            experience_level: experienceValue,
            category: categoryValue,
            in_progress: false,
          },
        );
      } else {
        alert('Please complete the entire form before submitting');
      }
    };

    postData = (data) => {
      const { cookies, toggleForms } = this.props;
      const headers = {
        'Content-Type': 'application/json',
        Authorization: `JWT ${cookies.get('auth_jwt')}`,
      };
      return axios.post('/api/v1/support-requests/', data, { headers })
        .then(() => {
          toggleForms();
        })
        .catch(() => {
          alert('Error: unable to reach server with help request. Please try again later.');
        });
    };

    render() {
      const {
        experienceValue, categoryValue, subjectValue, bodyValue,
      } = this.props;
      return (
        <div>
          <Card.Header>
            <b>
              Need help?
            </b>
          </Card.Header>
          <br />
          <label htmlFor="category">
            Select the issue you are having
            <Dropdown id="category" placeholder="Example: Logic Issues" fluid selection options={helpCategories} onChange={this.categorySelectChange} value={categoryValue} />
          </label>
          <br />
          <label htmlFor="experience-level">
            Select your experience level
            <Dropdown id="experience-level" placeholder="Example: Beginner" fluid selection options={experience} onChange={this.experienceSelectChange} value={experienceValue} />
          </label>
          <br />
          <label htmlFor="subject">
            Subject
            <Input fluid id="subject" placeholder="Describe your issue" onChange={this.subjectInputChange} />
          </label>
          <br />
          <br />
          <label htmlFor="body">
            More information
            <Input fluid id="body" placeholder="Provide any additional details" onChange={this.bodyInputChange} />
          </label>
          <br />
          <br />
          <Button className="ui primary button" onClick={() => this.validateForm(experienceValue, categoryValue, subjectValue, bodyValue)}>
            Submit
          </Button>
        </div>
      );
    }
}

ChatForm.propTypes = {
  bodyInputChange: PropTypes.func.isRequired,
  categorySelectChange: PropTypes.func.isRequired,
  subjectInputChange: PropTypes.func.isRequired,
  experienceSelectChange: PropTypes.func.isRequired,
  toggleForms: PropTypes.func.isRequired,
  bodyValue: PropTypes.string.isRequired,
  experienceValue: PropTypes.string.isRequired,
  subjectValue: PropTypes.string.isRequired,
  categoryValue: PropTypes.string.isRequired,
  cookies: PropTypes.instanceOf(Cookies).isRequired,
};

export default hot(module)(withCookies(connect(mapStateToProps, mapDispatchToProps)(ChatForm)));
