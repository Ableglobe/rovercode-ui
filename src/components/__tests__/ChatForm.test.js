import React from 'react';
import { Button } from 'semantic-ui-react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import configureStore from 'redux-mock-store';
import { Cookies } from 'react-cookie';
// import { stub } from 'sinon';
import ChatForm from '../ChatForm';

const cookiesValues = { auth_jwt: '1234' };
const cookies = new Cookies(cookiesValues);

describe('The ChatForm component', () => {
  const mockStore = configureStore();
  const context = { cookies };
  let store;

  beforeEach(() => {
    store = mockStore({
      chatform: {
        toggleForms: () => {},
        bodyValue: 'foobar',
        experienceValue: 'foobar',
        subjectValue: 'foobar',
        categoryValue: 'foobar',
      },
    });
    store.dispatch = jest.fn();
  });

  test('renders on the page with no errors', () => {
    const wrapper = mount(<ChatForm store={store} />, { context });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

//   beforeEach(() => {
//     store = mockStore({
//       chatform: {
//         //TODO Somethign here
//       },
//       chatapp: {
//         formHidden: false,
//         chatHidden: true,
//       },
//     });
//     store.dispatch = jest.fn();
//   });
//
// //   test('renders on the page with no errors', () => {
// //     const wrapper = mount(<ChatForm store={store} />);
// //     expect(toJson(wrapper)).toMatchSnapshot();
// //   });
//
//   test('toggles Chat', () => {
//     const toggleForms = stub();
//     const wrapper = shallow(<ChatApp store={store} toggleForms={toggleForms} />).dive();
//     wrapper.find("Button").prop('onClick')
//     expect(onClick.callCount).to.be.equal(1);
//     wrapper.find(Button).simulate('click');
//     expect(store.dispatch).toHaveBeenCalledWith(toggleForms());
//
//     // const handleClickStub = sinon.spy()
//     // const wrapper = mount(<Parent items={items} handleClick={handleClickStub} />)
//     // console.log(wrapper.state('clickedChild')) // prints false
//     // wrapper.find(ChildComponent).last().simulate('click')
//     // expect(handleClickStub.calledOnce).to.be.true // successful
//     // console.log(wrapper.state('clickedChild'))  // prints true
//   });
});