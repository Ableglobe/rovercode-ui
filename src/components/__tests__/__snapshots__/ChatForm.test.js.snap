// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The ChatForm component renders on the page with no errors 1`] = `
<withCookies(Component)
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(ChatForm)
    allCookies={Object {}}
    cookies={
      Cookies {
        "HAS_DOCUMENT_COOKIE": true,
        "changeListeners": Array [
          [Function],
        ],
        "cookies": Object {},
        "hooks": undefined,
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [MockFunction],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
  >
    <ChatForm
      allCookies={Object {}}
      bodyInputChange={[Function]}
      bodyValue="foobar"
      categorySelectChange={[Function]}
      categoryValue="foobar"
      cookies={
        Cookies {
          "HAS_DOCUMENT_COOKIE": true,
          "changeListeners": Array [
            [Function],
          ],
          "cookies": Object {},
          "hooks": undefined,
        }
      }
      experienceSelectChange={[Function]}
      experienceValue="foobar"
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [MockFunction],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
      storeSubscription={
        Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "clearActions": [Function],
            "dispatch": [MockFunction],
            "getActions": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
          },
          "unsubscribe": [Function],
        }
      }
      subjectInputChange={[Function]}
      subjectValue="foobar"
      toggleForms={[Function]}
    >
      <div>
        <CardHeader>
          <div
            className="header"
          >
            <b>
              Need help?
            </b>
          </div>
        </CardHeader>
        <br />
        <label
          htmlFor="category"
        >
          Select the issue you are having
          <Dropdown
            additionLabel="Add "
            additionPosition="top"
            closeOnBlur={true}
            deburr={false}
            fluid={true}
            icon="dropdown"
            id="category"
            minCharacters={1}
            noResultsMessage="No results found."
            onChange={[Function]}
            openOnFocus={true}
            options={
              Array [
                Object {
                  "text": "My code isn't doing what I think it should",
                  "value": "DEBUG",
                },
                Object {
                  "text": "I don't know what to do next",
                  "value": "WHAT_NEXT",
                },
                Object {
                  "text": "I'm looking for fun ideas",
                  "value": "FUN_IDEAS",
                },
              ]
            }
            placeholder="Example: Logic Issues"
            renderLabel={[Function]}
            searchInput="text"
            selectOnBlur={true}
            selectOnNavigation={true}
            selection={true}
            value="foobar"
            wrapSelection={true}
          >
            <div
              aria-expanded={false}
              className="ui fluid selection dropdown"
              id="category"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              role="listbox"
              tabIndex={0}
            >
              <div
                aria-live="polite"
                className="text"
                role="alert"
              />
              <Icon
                as="i"
                name="dropdown"
                onClick={[Function]}
              >
                <i
                  aria-hidden="true"
                  className="dropdown icon"
                  onClick={[Function]}
                />
              </Icon>
              <DropdownMenu>
                <div
                  className="menu transition"
                >
                  <DropdownItem
                    active={false}
                    key="DEBUG"
                    onClick={[Function]}
                    selected={true}
                    style={
                      Object {
                        "pointerEvents": "all",
                      }
                    }
                    text="My code isn't doing what I think it should"
                    value="DEBUG"
                  >
                    <div
                      aria-checked={false}
                      aria-selected={true}
                      className="selected item"
                      onClick={[Function]}
                      role="option"
                      style={
                        Object {
                          "pointerEvents": "all",
                        }
                      }
                    >
                      <span
                        className="text"
                      >
                        My code isn't doing what I think it should
                      </span>
                    </div>
                  </DropdownItem>
                  <DropdownItem
                    active={false}
                    key="WHAT_NEXT"
                    onClick={[Function]}
                    selected={false}
                    style={
                      Object {
                        "pointerEvents": "all",
                      }
                    }
                    text="I don't know what to do next"
                    value="WHAT_NEXT"
                  >
                    <div
                      aria-checked={false}
                      aria-selected={false}
                      className="item"
                      onClick={[Function]}
                      role="option"
                      style={
                        Object {
                          "pointerEvents": "all",
                        }
                      }
                    >
                      <span
                        className="text"
                      >
                        I don't know what to do next
                      </span>
                    </div>
                  </DropdownItem>
                  <DropdownItem
                    active={false}
                    key="FUN_IDEAS"
                    onClick={[Function]}
                    selected={false}
                    style={
                      Object {
                        "pointerEvents": "all",
                      }
                    }
                    text="I'm looking for fun ideas"
                    value="FUN_IDEAS"
                  >
                    <div
                      aria-checked={false}
                      aria-selected={false}
                      className="item"
                      onClick={[Function]}
                      role="option"
                      style={
                        Object {
                          "pointerEvents": "all",
                        }
                      }
                    >
                      <span
                        className="text"
                      >
                        I'm looking for fun ideas
                      </span>
                    </div>
                  </DropdownItem>
                </div>
              </DropdownMenu>
            </div>
          </Dropdown>
        </label>
        <br />
        <label
          htmlFor="experience-level"
        >
          Select your experience level
          <Dropdown
            additionLabel="Add "
            additionPosition="top"
            closeOnBlur={true}
            deburr={false}
            fluid={true}
            icon="dropdown"
            id="experience-level"
            minCharacters={1}
            noResultsMessage="No results found."
            onChange={[Function]}
            openOnFocus={true}
            options={
              Array [
                Object {
                  "text": "This is my first time on Rovercode",
                  "value": 0,
                },
                Object {
                  "text": "I'm pretty new to Rovercode",
                  "value": 1,
                },
                Object {
                  "text": "I've used Rovercode for a while",
                  "value": 2,
                },
                Object {
                  "text": "I'm a Rovercode expert",
                  "value": 3,
                },
              ]
            }
            placeholder="Example: Beginner"
            renderLabel={[Function]}
            searchInput="text"
            selectOnBlur={true}
            selectOnNavigation={true}
            selection={true}
            value="foobar"
            wrapSelection={true}
          >
            <div
              aria-expanded={false}
              className="ui fluid selection dropdown"
              id="experience-level"
              onBlur={[Function]}
              onChange={[Function]}
              onClick={[Function]}
              onFocus={[Function]}
              onMouseDown={[Function]}
              role="listbox"
              tabIndex={0}
            >
              <div
                aria-live="polite"
                className="text"
                role="alert"
              />
              <Icon
                as="i"
                name="dropdown"
                onClick={[Function]}
              >
                <i
                  aria-hidden="true"
                  className="dropdown icon"
                  onClick={[Function]}
                />
              </Icon>
              <DropdownMenu>
                <div
                  className="menu transition"
                >
                  <DropdownItem
                    active={false}
                    key="0"
                    onClick={[Function]}
                    selected={true}
                    style={
                      Object {
                        "pointerEvents": "all",
                      }
                    }
                    text="This is my first time on Rovercode"
                    value={0}
                  >
                    <div
                      aria-checked={false}
                      aria-selected={true}
                      className="selected item"
                      onClick={[Function]}
                      role="option"
                      style={
                        Object {
                          "pointerEvents": "all",
                        }
                      }
                    >
                      <span
                        className="text"
                      >
                        This is my first time on Rovercode
                      </span>
                    </div>
                  </DropdownItem>
                  <DropdownItem
                    active={false}
                    key="1"
                    onClick={[Function]}
                    selected={false}
                    style={
                      Object {
                        "pointerEvents": "all",
                      }
                    }
                    text="I'm pretty new to Rovercode"
                    value={1}
                  >
                    <div
                      aria-checked={false}
                      aria-selected={false}
                      className="item"
                      onClick={[Function]}
                      role="option"
                      style={
                        Object {
                          "pointerEvents": "all",
                        }
                      }
                    >
                      <span
                        className="text"
                      >
                        I'm pretty new to Rovercode
                      </span>
                    </div>
                  </DropdownItem>
                  <DropdownItem
                    active={false}
                    key="2"
                    onClick={[Function]}
                    selected={false}
                    style={
                      Object {
                        "pointerEvents": "all",
                      }
                    }
                    text="I've used Rovercode for a while"
                    value={2}
                  >
                    <div
                      aria-checked={false}
                      aria-selected={false}
                      className="item"
                      onClick={[Function]}
                      role="option"
                      style={
                        Object {
                          "pointerEvents": "all",
                        }
                      }
                    >
                      <span
                        className="text"
                      >
                        I've used Rovercode for a while
                      </span>
                    </div>
                  </DropdownItem>
                  <DropdownItem
                    active={false}
                    key="3"
                    onClick={[Function]}
                    selected={false}
                    style={
                      Object {
                        "pointerEvents": "all",
                      }
                    }
                    text="I'm a Rovercode expert"
                    value={3}
                  >
                    <div
                      aria-checked={false}
                      aria-selected={false}
                      className="item"
                      onClick={[Function]}
                      role="option"
                      style={
                        Object {
                          "pointerEvents": "all",
                        }
                      }
                    >
                      <span
                        className="text"
                      >
                        I'm a Rovercode expert
                      </span>
                    </div>
                  </DropdownItem>
                </div>
              </DropdownMenu>
            </div>
          </Dropdown>
        </label>
        <br />
        <label
          htmlFor="subject"
        >
          Subject
          <Input
            fluid={true}
            id="subject"
            onChange={[Function]}
            placeholder="Describe your issue"
            type="text"
          >
            <div
              className="ui fluid input"
            >
              <input
                id="subject"
                onChange={[Function]}
                placeholder="Describe your issue"
                type="text"
              />
            </div>
          </Input>
        </label>
        <br />
        <br />
        <label
          htmlFor="body"
        >
          More information
          <Input
            fluid={true}
            id="body"
            onChange={[Function]}
            placeholder="Provide any additional details"
            type="text"
          >
            <div
              className="ui fluid input"
            >
              <input
                id="body"
                onChange={[Function]}
                placeholder="Provide any additional details"
                type="text"
              />
            </div>
          </Input>
        </label>
        <br />
        <br />
        <Button
          as="button"
          className="ui primary button"
          onClick={[Function]}
          role="button"
        >
          <button
            className="ui button ui primary button"
            onClick={[Function]}
            role="button"
          >
            Submit
          </button>
        </Button>
      </div>
    </ChatForm>
  </Connect(ChatForm)>
</withCookies(Component)>
`;
